File: tp-8/Nave/haskell/nave.hs
module Empleado (
    Nave, construir, ingresarT
    )
where

import Map
import MaxHeap

data Nave = N (Map SectorId Sector) (Map Nombre Tripulante) (MaxHeap Tripulante) deriving (Show, Eq)


{-
    INVARIANTE DE REPRENSENTACION:  
    El tipo  SectorId identifica a un solo sector y no se repiten en mas de un tipo SectorId
    El tipo Nombre identifica a un solo Tripulante y no se repiten en mas de un tipo Tripulante
    Los tipo Tripulantes no necesariamente existen en el tipo Sector  
    En el MaxHeap los tripulantes no se repiten y son los mismos que en el Map de tripulantes
-}

// costo: O(S), lineal por addSectores que tiene un costo O(S), siendo S el largo de la lista del tipo SectorId
construir:: [SectorId] -> Nave
construir ss = (N (addSectores ss emptyM) emptyM emptyH)

// costo:  O(S), lineal, siendo S el largo de la lista del tipo SectorId
addSectores:: [SectorId] -> Map SectorId Sector -> Map SectorId Sector 
addSectores [] = emptyM
addSectores (x:xs) = assocM x (addSectores xs) 

// costo: O(log T), siendo T el valor de K + M, donde assocM tiene un costo O(log k) siendo k las keys del map de SectorId y sumado a que insertH tiene un costo de O(log M) siendo M el largo logaritmico
ingresarT:: Nombre -> Rango -> Nave -> Nave
ingresarT n r (N ss nt ts) = 
    let t = (creatT n r) in (N ss (assocM n t nt) (insertH t ts) 

// costo: O(log M), siendo M el valor de K, lookupM tiene un costo O(log k) siendo k la cantidad de keys del map y sumado a sectoresT tiene un costo de constante O(1)
sectoresAsignados:: Nombre -> Nave -> Set SectorId
sectoresAsignados n (N ss nt ts) = case lookupM n nt on 
Just t -> sectoresT t
Nothing -> error "el Nombre debe  existir en la nave"

// costo: O(log S), siendo S la suma de K + funciones constantes que suman pero no lo suficiente para modificar el costo principal, lookupM teniendo un costo de O(log K), tripulantesS y ComponenteS ambas siendo constantes, no modifican en gran medida los costos en general
datosDeSector : SectorId -> Nave -> (Set Nombre, [Componente])
datosDeSector sId (N ss nt ts) =  
    let s = lookupM sId ss in ((tripulantesS s), (ComponenteS s))  

File: tp-8/Nave/cplusplus/nave/Nave.cpp
#include <iostream>
#include "Nave.h"
#include "../map/Map.h"
#include "../set/Set.h"
#include "../maxHeap/MaxHeap.h"
#include "../shared.h"
using namespace std;

//costo: constante O(1)
Nave construir (SectorId ss){
    NaveSt* n = new NaveSt;

    tripulantes = emptyM();
    sectores = emptyM();
    OrdRangoTripulantes = emptyH();

    n->tripulantes = tripulantes;
    n->sectores = sectores;
    n->OrdRangoTripulantes= OrdRangoTripulantes;
    return n;
};

//costo: O(n) donde n es la cantidad de tripulantes en la nave
void ingresarT(Nombre n, Rango r, Nave n){
    Tripulante t = crearT(n, r);
    assocM(n, t, n->tripulantes); 
    insertH(t, n->OrdRangoTripulantes);
};

//costo: O(n) donde n es la cantidad de tripulantes en la nave
Set sectoresAsignados (Nombre n, Nave n){
    Tripulante t = lookupM(n, n->tripulantes);

    if(t != nullptr){
       return sectoresT(t);
    }
    return emptyS();
};
 
//costo: O(n) donde n es la cantidad de sectores en la nave
pair<Set, Componente*> datosDeSector(SectorId sId, Nave n){
    Sector s = lookupM(sId, n->sectores);
    if(s != nullptr){
        return make_pair(sectoresC(s), componentesC(s));
    }
    return make_pair(emptyS(), nullptr);
};


//costo: O(n) donde n es la cantidad de tripulantes en la nave
Tripulante* tripulantesN (Nave n, int cant){
    string[cant] keys = domM(n->tripulantes);
    int i = 0;
    while(i < cant){
        keys[i]= lookupM(n->tripulantes[i], n->tripulantes);
        i++;
    }
    return keys;
};

//costo: O(n) donde n es la cantidad de componentes a agregar
void agregarASector (Componentes* cs, int cant, SectorId sId, Nave n){
    int i = 0;
    Sector s = lookupM(sId, n->sectores);
    while(i < cant){
        agregarC(cs[i], s);
        i++;
    }
    delete s;
    assocM(sId, s, n->sectores);
};

//costo: O(n) donde n es la cantidad de tripulantes en la nave
void asignarASector(Nombre n, SectorId sId, Nave n){
    Sector s = lookupM(sId, n->sectores);
    agregarT(n, s);
    assocM(sId, s, n->sectores);
    delete s;
};

int testNave(){
    return 0;
}